Source/Adobe.Target.Client/OnDevice/DecisioningDetailsExecutor.cs

```diff
@@ -0,0 +1,323 @@
+            ExecuteResponse executeResponse,
+            IList<Notification> notifications)
+        {
+            var rules = GetDetailsRules(details, ruleSet);
+            if (rules == null)
+            {
+                UnhandledResponse(details, prefetchResponse, executeResponse);
+                return;
+            }
+
+            var propertyToken = deliveryRequest.DeliveryRequest.Property?.Token;
+            var handledAtLeastOnce = false;
+            var skipKeySet = new HashSet<string>();
+
+            foreach (var rule in rules)
+            {
+                if (PropertyTokenMismatch(rule.PropertyTokens, propertyToken))
+                {
+                    continue;
+                }
+
+                if (rule.RuleKey != null && skipKeySet.Contains(rule.RuleKey))
+                {
+                    continue;
+                }
+
+                var consequence = this.ruleExecutor.ExecuteRule(localContext, details, visitorId, rule, responseTokens);
+                if (!HandleResult(consequence, rule, details, prefetchResponse, executeResponse, notifications, ruleSet.GlobalMbox))
+                {
+                    continue;
+                }
+
+                handledAtLeastOnce = true;
+                if (rule.RuleKey != null)
+                {
+                    skipKeySet.Add(rule.RuleKey);
+                }
+
+                if (details.Match(
+                    _ => false,
+                    mboxRequest => true,
+                    _ => false))
+                {
+                    break;
+                }
+            }
+
+            if (!handledAtLeastOnce)
+            {
+                UnhandledResponse(details, prefetchResponse, executeResponse);
+            }
+        }
+
+        private static bool HandleResult(
+            IDictionary<string, object> consequence,
+            OnDeviceDecisioningRule rule,
+            RequestDetailsUnion details,
+            PrefetchResponse prefetchResponse,
+            ExecuteResponse executeResponse,
+            IList<Notification> notifications,
+            string globalMbox = "target-global-mbox")
+        {
+            if (consequence == null || consequence.Count == 0)
+            {
+                return false;
+            }
+
+            consequence.TryGetValue(Name, out var nameObject);
+            var consequenceName = (string)nameObject;
+
+            consequence.TryGetValue(Options, out var optionsObject);
+            var consequenceOptions = (List<Option>)optionsObject;
+
+            consequence.TryGetValue(Metrics, out var metricsObject);
+            var consequenceMetrics = GetMetrics(metricsObject);
+
+            if (executeResponse != null)
+            {
+                var notification = CreateNotification(details, consequenceOptions, globalMbox);
+                notifications.Add(notification);
+            }
+
+            return details.Match(
+                _ =>
```

Comment:
Could we please extract every match condition to a separate method. `HandleResult` is a pretty big method.

Reply:
Fixed